File: research/measurement/sdt.py
"""
SIGNAL DETECTION THEORY, AS PUBLISHED, WITH THE CORRECTION NAMED.

F5 is the falsification that says an apparent improvement in discrimination is a shift in
response criterion: a player who simply captures more often raises the hit rate AND the
false-alarm rate, and accuracy alone cannot tell the two apart. The defence is not to argue that
this will not happen. It is to never report accuracy without the two rates that produced it.

FORMULAE ARE FROM Stanislaw, H., & Todorov, N. (1999), "Calculation of signal detection theory
measures", Behavior Research Methods, Instruments, & Computers 31(1), 137-149. Nothing here is
invented; `tests/research/sdt.test.ts` pins the worked examples from that paper's tables so a
later edit that "simplifies" one of these formulae goes red.

WHAT THIS FILE CANNOT DO, stated here because it is the thing most likely to be forgotten:
d' does not repair a confounded item set. If T+ items are systematically easier than T- items,
then a participant with no discrimination ability whatsoever will still produce a hit rate above
its false-alarm rate, and d' will be positive and stable and completely uninterpretable.
`negative_controls.py::item_difficulty_confound` exists to make that failure visible.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, asdict

from scipy.stats import norm

SDT_VERSION = "1.0.0"


@dataclass(frozen=True)
class Counts:
    hits: int
    misses: int
    false_alarms: int
    correct_rejections: int

    @property
    def signal_trials(self) -> int:
        return self.hits + self.misses

    @property
    def noise_trials(self) -> int:
        return self.false_alarms + self.correct_rejections


@dataclass(frozen=True)
class Sdt:
    hit_rate: float
    false_alarm_rate: float
    d_prime: float
    criterion_c: float
    beta: float
    a_prime: float
    b_double_prime_d: float
    correction: str
    signal_trials: int
    noise_trials: int
    hits: int
    false_alarms: int

    def as_dict(self) -> dict:
        return asdict(self)


def _loglinear(counts: Counts) -> tuple[float, float]:
    """
    Hautus (1995) loglinear correction: add 0.5 to hits and false alarms, 1 to each trial total.

    APPLIED ALWAYS, NOT ONLY AT THE EXTREMES, and that choice matters. Correcting only the cells
    that came out 0 or 1 makes the correction itself depend on the data, which biases the
    corrected estimates in a direction that varies with the true rate. Hautus's comparison of
    corrections is the reason this is the default rather than the 1/(2N) rule; applying it
    uniformly is what that recommendation actually is.
    """
    h = (counts.hits + 0.5) / (counts.signal_trials + 1.0)
    f = (counts.false_alarms + 0.5) / (counts.noise_trials + 1.0)
    return h, f


def compute(counts: Counts, correction: str = "loglinear") -> Sdt:
    """
    d', c, beta, A' and B''_D from one 2x2 table.

    `correction="none"` is available and will raise on a degenerate table rather than return
    infinity, because an infinite d' reported as a number is how a ceiling effect gets published
    as a large effect.
    """
    if correction == "loglinear":
        h, f = _loglinear(counts)
    elif correction == "none":
        if counts.signal_trials == 0 or counts.noise_trials == 0:
            raise ValueError("no trials of one type; d' is undefined, not infinite")
        h = counts.hits / counts.signal_trials
        f = counts.false_alarms / counts.noise_trials
        if h in (0.0, 1.0) or f in (0.0, 1.0):
            raise ValueError(
                "extreme rate with correction='none'; z(0) and z(1) are infinite. "
                "Choose a correction and say which one you chose."
            )
    else:
        raise ValueError(f"unknown correction {correction!r}")

    zh, zf = norm.ppf(h), norm.ppf(f)
    d = zh - zf
    c = -0.5 * (zh + zf)
    # beta as the likelihood ratio at the criterion (S&T eq. for the equal-variance model).
    beta = math.exp((zf * zf - zh * zh) / 2.0)

    # A' and B''_D: the nonparametric pair. Reported BESIDE d', never instead of it, because
    # they make a different (weaker) assumption and can disagree -- and when they disagree, the
    # equal-variance assumption is what is being called into question.
    if h >= f:
        a_prime = 0.5 + ((h - f) * (1 + h - f)) / (4 * h * (1 - f))
    else:
        a_prime = 0.5 - ((f - h) * (1 + f - h)) / (4 * f * (1 - h))

    num = h * (1 - h) - f * (1 - f)
    den = h * (1 - h) + f * (1 - f)
    b_dd = num / den if den != 0 else 0.0

    return Sdt(
        hit_rate=h,
        false_alarm_rate=f,
        d_prime=d,
        criterion_c=c,
        beta=beta,
        a_prime=a_prime,
        b_double_prime_d=b_dd,
        correction=correction,
        signal_trials=counts.signal_trials,
        noise_trials=counts.noise_trials,
        hits=counts.hits,
        false_alarms=counts.false_alarms,
    )


def from_trials(trials) -> Counts:
    """
    Fold an iterable of (trigger_state, behaviour) into a 2x2 table.

    `trigger_state` must be "positive" or "negative" and `behaviour` must be 0 or 1. Anything
    else raises. An UNKNOWN item reaching a scoring function is a bug in the item bank, and the
    right response to it is a crash rather than a quietly smaller denominator.
    """
    h = m = fa = cr = 0
    for state, behaviour in trials:
        if behaviour not in (0, 1):
            raise ValueError(f"behaviour must be 0 or 1, got {behaviour!r}")
        if state == "positive":
            h += behaviour
            m += 1 - behaviour
        elif state == "negative":
            fa += behaviour
            cr += 1 - behaviour
        else:
            raise ValueError(
                f"trigger_state {state!r} may not be scored; UNKNOWN items are excluded "
                "by the item bank, not by the scorer"
            )
    return Counts(hits=h, misses=m, false_alarms=fa, correct_rejections=cr)


def wilson_interval(successes: int, trials: int, z: float = 1.96) -> tuple[float, float]:
    """
    Wilson score interval for a proportion.

    WILSON RATHER THAN WALD because the rates here go near 0 and 1 by design -- a strong player's
    false-alarm rate is supposed to be small -- and the Wald interval is known to have coverage
    far below nominal there and to produce bounds outside [0, 1]. Brown, Cai & DasGupta (2001).
    """
    if trials == 0:
        return (float("nan"), float("nan"))
    p = successes / trials
    denom = 1 + z * z / trials
    centre = (p + z * z / (2 * trials)) / denom
    half = (z * math.sqrt(p * (1 - p) / trials + z * z / (4 * trials * trials))) / denom
    return (max(0.0, centre - half), min(1.0, centre + half))


def standardized_mean_difference(a, b) -> float:
    """
    Cohen's d on two samples, used for F2's covariate balance rather than for an effect.

    A convention this program does NOT adopt: that |d| < 0.1 means "balanced". That number is a
    what-works-clearinghouse baseline-equivalence convention for a different design; whether it
    is the right bar here has not been established, so the number is REPORTED and the reader is
    told what it is not.
    """
    import statistics

    a, b = list(a), list(b)
    if len(a) < 2 or len(b) < 2:
        return float("nan")
    va, vb = statistics.variance(a), statistics.variance(b)
    pooled = math.sqrt(((len(a) - 1) * va + (len(b) - 1) * vb) / (len(a) + len(b) - 2))
    if pooled == 0:
        return 0.0
    return (statistics.mean(a) - statistics.mean(b)) / pooled

File: research/measurement/sdt.ts
/**
 * SIGNAL DETECTION THEORY IN TYPESCRIPT, PORTED AFTER EQUIVALENCE.
 *
 * `research/measurement/sdt.py` is the implementation every corpus number in
 * `docs/measurement/FALSIFICATION_REGISTER.md` was produced by. This is the same arithmetic in
 * the language the negative controls and their tests run in, and it exists for one reason: the
 * controls have to run inside `npm test`, on every build, with no Python in the loop. A control
 * that only runs when somebody remembers to run it is not a control.
 *
 * THIS IS `PORT_AFTER_EQUIVALENCE` IN THE SENSE `docs/decisions/README.md` DEFINES. The port is
 * differenced against the original before it is used: `equivalence/sdt_grid.json` is generated
 * BY the Python and `tests/research/measurement-sdt.test.ts` asserts this file reproduces every
 * row of it. If the two ever disagree the build goes red, which is the only thing that makes two
 * implementations of one formula safe to have.
 *
 * FORMULAE: Stanislaw, H., & Todorov, N. (1999), Behavior Research Methods, Instruments, &
 * Computers 31(1), 137-149. Loglinear correction: Hautus, M. J. (1995), BRMIC 27(1), 46-51.
 *
 * NOT IN THE PRODUCT BUNDLE. This file lives under `research/` and nothing in `client/`,
 * `server/` or `shared/` imports it. Whether any of it earns its way into the product is a
 * question `docs/measurement/GO_NO_GO.md` answers, and the answer there is not yet.
 */

export interface Counts {
  readonly hits: number;
  readonly misses: number;
  readonly falseAlarms: number;
  readonly correctRejections: number;
}

export interface SdtResult {
  readonly hitRate: number;
  readonly falseAlarmRate: number;
  readonly dPrime: number;
  readonly criterionC: number;
  readonly beta: number;
  readonly aPrime: number;
  readonly bDoublePrimeD: number;
  readonly correction: "loglinear" | "none";
  readonly signalTrials: number;
  readonly noiseTrials: number;
}

/**
 * The inverse standard normal CDF.
 *
 * Acklam's rational approximation, whose stated relative error is below 1.15e-9 over the whole
 * open interval. NO REFINEMENT STEP, and the reason is worth stating: a Halley step needs an
 * `erfc` of its own, a second approximation with its own error, and the honest way to describe
 * the result would then be "two approximations, one tolerance". The equivalence fixture is
 * asserted at 1e-8 absolute, which this clears and which is four orders of magnitude finer than
 * any d' this program will ever interpret.
 */
export function normalQuantile(p: number): number {
  if (!(p > 0 && p < 1)) {
    throw new RangeError(
      `normalQuantile needs 0 < p < 1, got ${p}. An extreme rate reached this function, ` +
        `which means a correction was skipped upstream.`,
    );
  }
  const a = [
    -3.969683028665376e1, 2.209460984245205e2, -2.759285104469687e2,
    1.38357751867269e2, -3.066479806614716e1, 2.506628277459239,
  ];
  const b = [
    -5.447609879822406e1, 1.615858368580409e2, -1.556989798598866e2,
    6.680131188771972e1, -1.328068155288572e1,
  ];
  const c = [
    -7.784894002430293e-3, -3.223964580411365e-1, -2.400758277161838,
    -2.549732539343734, 4.374664141464968, 2.938163982698783,
  ];
  const d = [
    7.784695709041462e-3, 3.224671290700398e-1, 2.445134137142996,
    3.754408661907416,
  ];
  const pLow = 0.02425;
  const pHigh = 1 - pLow;

  if (p < pLow) {
    const q = Math.sqrt(-2 * Math.log(p));
    return (
      (((((c[0] * q + c[1]) * q + c[2]) * q + c[3]) * q + c[4]) * q + c[5]) /
      ((((d[0] * q + d[1]) * q + d[2]) * q + d[3]) * q + 1)
    );
  }
  if (p > pHigh) {
    const q = Math.sqrt(-2 * Math.log(1 - p));
    return (
      -(((((c[0] * q + c[1]) * q + c[2]) * q + c[3]) * q + c[4]) * q + c[5]) /
      ((((d[0] * q + d[1]) * q + d[2]) * q + d[3]) * q + 1)
    );
  }
  const q = p - 0.5;
  const r = q * q;
  return (
    ((((((a[0] * r + a[1]) * r + a[2]) * r + a[3]) * r + a[4]) * r + a[5]) * q) /
    (((((b[0] * r + b[1]) * r + b[2]) * r + b[3]) * r + b[4]) * r + 1)
  );
}

/**
 * Hits+0.5 over signal trials+1, false alarms+0.5 over noise trials+1.
 *
 * APPLIED TO EVERY TABLE, not only to the degenerate ones, because a correction whose
 * application depends on the data biases the corrected estimate by an amount that varies with
 * the true rate. This is what Hautus's recommendation actually is, and the alternative --
 * correcting only zeros and ones -- is the version that is easy to write and wrong.
 */
export function loglinearRates(counts: Counts): { hitRate: number; falseAlarmRate: number } {
  const signal = counts.hits + counts.misses;
  const noise = counts.falseAlarms + counts.correctRejections;
  return {
    hitRate: (counts.hits + 0.5) / (signal + 1),
    falseAlarmRate: (counts.falseAlarms + 0.5) / (noise + 1),
  };
}

export function computeSdt(counts: Counts, correction: "loglinear" | "none" = "loglinear"): SdtResult {
  const signalTrials = counts.hits + counts.misses;
  const noiseTrials = counts.falseAlarms + counts.correctRejections;

  let hitRate: number;
  let falseAlarmRate: number;
  if (correction === "loglinear") {
    ({ hitRate, falseAlarmRate } = loglinearRates(counts));
  } else {
    if (signalTrials === 0 || noiseTrials === 0) {
      throw new RangeError("no trials of one type; d' is undefined, not infinite");
    }
    hitRate = counts.hits / signalTrials;
    falseAlarmRate = counts.falseAlarms / noiseTrials;
  }

  const zh = normalQuantile(hitRate);
  const zf = normalQuantile(falseAlarmRate);
  const dPrime = zh - zf;
  const criterionC = -0.5 * (zh + zf);
  const beta = Math.exp((zf * zf - zh * zh) / 2);

  const h = hitRate;
  const f = falseAlarmRate;
  const aPrime =
    h >= f
      ? 0.5 + ((h - f) * (1 + h - f)) / (4 * h * (1 - f))
      : 0.5 - ((f - h) * (1 + f - h)) / (4 * f * (1 - h));

  const num = h * (1 - h) - f * (1 - f);
  const den = h * (1 - h) + f * (1 - f);
  const bDoublePrimeD = den === 0 ? 0 : num / den;

  return {
    hitRate,
    falseAlarmRate,
    dPrime,
    criterionC,
    beta,
    aPrime,
    bDoublePrimeD,
    correction,
    signalTrials,
    noiseTrials,
  };
}

export type TriggerState = "positive" | "negative" | "unknown";

export interface Trial {
  readonly triggerState: TriggerState;
  readonly behaviour: 0 | 1;
}

/**
 * Fold trials into a 2x2 table.
 *
 * AN `unknown` ITEM THROWS. It does not shrink a denominator quietly. The item bank is what
 * excludes UNKNOWNs; a scorer that tolerated them would let an exclusion rule be decided by
 * whichever items happened to arrive.
 */
export function tabulate(trials: readonly Trial[]): Counts {
  let hits = 0;
  let misses = 0;
  let falseAlarms = 0;
  let correctRejections = 0;
  for (const t of trials) {
    if (t.triggerState === "positive") {
      if (t.behaviour === 1) hits += 1;
      else misses += 1;
    } else if (t.triggerState === "negative") {
      if (t.behaviour === 1) falseAlarms += 1;
      else correctRejections += 1;
    } else {
      throw new RangeError(
        "an UNKNOWN item reached the scorer; UNKNOWNs are excluded by the item bank, " +
          "never by the scoring function",
      );
    }
  }
  return { hits, misses, falseAlarms, correctRejections };
}

/** Wilson score interval. Wald is not used: these rates sit near 0 and 1 by design. */
export function wilsonInterval(successes: number, trials: number, z = 1.96): [number, number] {
  if (trials === 0) return [Number.NaN, Number.NaN];
  const p = successes / trials;
  const denom = 1 + (z * z) / trials;
  const centre = (p + (z * z) / (2 * trials)) / denom;
  const half =
    (z * Math.sqrt((p * (1 - p)) / trials + (z * z) / (4 * trials * trials))) / denom;
  return [Math.max(0, centre - half), Math.min(1, centre + half)];
}
